app/src/main/java/com/keylesspalace/tusky/components/trending/TrendingPagingAdapter.kt

```diff
@@ -0,0 +1,147 @@
+                TrendingTagViewHolder(binding)
+            }
+
+            else -> {
+                throw IllegalStateException("Trending view cannot be created due to viewType mismatch.")
+            }
+        }
+    }
+
+    override fun onBindViewHolder(viewHolder: RecyclerView.ViewHolder, position: Int) {
+        bindViewHolder(viewHolder, position, null)
+    }
+
+    override fun onBindViewHolder(
+        viewHolder: RecyclerView.ViewHolder,
+        position: Int,
+        payloads: List<*>
+    ) {
+        bindViewHolder(viewHolder, position, payloads)
+    }
+
+    private fun bindViewHolder(
+        viewHolder: RecyclerView.ViewHolder,
+        position: Int,
+        payloads: List<*>?
+    ) {
+        when (val header = getItem(position)) {
+            is TrendingViewData.Tag -> {
+                val maxTrendingValue = currentList
+                    .flatMap { trendingViewData ->
+                        trendingViewData.asTagOrNull()?.tag?.history ?: emptyList()
+                    }
+                    .mapNotNull { it.uses.toLongOrNull() }
+                    .maxOrNull() ?: 1
+
+                val holder = viewHolder as TrendingTagViewHolder
+                holder.setup(header, maxTrendingValue, trendingListener)
+            }
+
+            is TrendingViewData.Header -> {
+                val holder = viewHolder as TrendingDateViewHolder
+                holder.setup(header.start, header.end)
+            }
+        }
+    }
+
+    override fun getItemViewType(position: Int): Int {
+        return if (getItem(position) is TrendingViewData.Tag) {
+            VIEW_TYPE_TAG
+        } else if (getItem(position) is TrendingViewData.Header) {
+            VIEW_TYPE_HEADER
+        } else {
+            throw IllegalStateException("Trending view cannot be retrieved due to viewType mismatch.")
+        }
+    }
+
+    companion object {
+        const val VIEW_TYPE_HEADER = 0
+        const val VIEW_TYPE_TAG = 1
+
+        val TrendingDifferCallback = object : DiffUtil.ItemCallback<TrendingViewData>() {
+            override fun areItemsTheSame(
+                oldItem: TrendingViewData,
+                newItem: TrendingViewData
+            ): Boolean {
+                return oldItem.id == newItem.id
+            }
+
+            override fun areContentsTheSame(
+                oldItem: TrendingViewData,
+                newItem: TrendingViewData
+            ): Boolean {
+                return false // Items are different always. It allows to refresh timestamp on every view holder update
+            }
+
+            override fun getChangePayload(
+                oldItem: TrendingViewData,
+                newItem: TrendingViewData
+            ): Any? {
+                return if (oldItem == newItem) {
+                    // If items are equal - update timestamp only
+                    listOf(StatusBaseViewHolder.Key.KEY_CREATED)
```

Comment:
There are no timestamps shown here, you can scrap the whole change payload business